### PR TITLE
Migrate to halo2 version with `AssignedCell`

### PR DESCRIPTION
**Pull request by str4d**
Best reviewed by commit.

### CODE REVIEW

File: src/circuit/gadget/poseidon.rs [hunk 1]
```diff
@@ -144,7 +145,13 @@ impl<
         chip.initial_state(&mut layouter, &domain)
             .map(|state| Duplex {
                 chip,
-                sponge: Sponge::Absorbing([None; RATE]),
+                sponge: Sponge::Absorbing(
+                    (0..RATE)
+                        .map(|_| None)
+                        .collect::<Vec<_>>()
+                        .try_into()
+                        .unwrap(),
```
Comment (daira) [inline src/circuit/gadget/poseidon.rs:152]: Why is this change needed?

File: src/circuit/gadget/ecc/chip/add.rs [hunk 2]
```diff
@@ -310,11 +310,11 @@ impl Config {
...
                         if x_p.is_zero_vartime() {
                             // 0 + Q = Q
-                            (x_q, y_q)
+                            (*x_q, *y_q)
                         } else if x_q.is_zero_vartime() {
                             // P + 0 = P
-                            (x_p, y_p)
-                        } else if (x_q == x_p) && (y_q == -y_p) {
+                            (*x_p, *y_p)
+                        } else if (x_q == x_p) && (*y_q == -y_p) {
```
Comment (daira) [inline src/circuit/gadget/ecc/chip/add.rs:317]: Why does only `y_q` change here?

File: src/circuit/gadget/ecc/chip/add_incomplete.rs [hunk 3]
```diff
@@ -136,8 +136,8 @@ impl Config {
         )?;
 
         let result = NonIdentityEccPoint {
-            x: CellValue::<pallas::Base>::new(x_r_var, x_r),
-            y: CellValue::<pallas::Base>::new(y_r_var, y_r),
```
Comment (daira) [inline src/circuit/gadget/ecc/chip/add_incomplete.rs:139]: Consider renaming these to `x_r_cell` and `y_r_cell` for consistency.

File: src/circuit/gadget/ecc/chip/add_incomplete.rs [hunk 4]
```diff
@@ -136,8 +136,8 @@ impl Config {
...
+            x: x_r_var,
+            y: y_r_var,
```
Comment (daira) [inline src/circuit/gadget/ecc/chip/add_incomplete.rs:140]: Consider renaming these to `x_r_cell` and `y_r_cell` for consistency.

File: src/circuit/gadget/ecc/chip/mul_fixed/short.rs [hunk 5]
```diff
@@ -183,7 +183,7 @@ impl Config {
 
                 Ok(EccPoint {
                     x: magnitude_mul.x,
-                    y: CellValue::new(y_var, y_val),
+                    y: y_var,
```
Comment (daira) [inline src/circuit/gadget/ecc/chip/mul_fixed/short.rs:183]: Consider renaming this to `y_cell` for consistency.

File: src/circuit/gadget/ecc/chip/witness_point.rs [hunk 6]
```diff
@@ -81,10 +81,7 @@ impl Config {
         let y_var =
             region.assign_advice(|| "y", self.y, offset, || y_val.ok_or(Error::Synthesis))?;
 
-        Ok((
-            CellValue::<pallas::Base>::new(x_var, x_val),
-            CellValue::<pallas::Base>::new(y_var, y_val),
-        ))
+        Ok((x_var, y_var))
```
Comment (daira) [inline src/circuit/gadget/ecc/chip/witness_point.rs:90]: Consider renaming these to `x_cell` and `y_cell`.

File: src/circuit/gadget/sinsemilla/chip/hash_to_point.rs [hunk 7]
```diff
@@ -46,11 +46,8 @@ impl SinsemillaChip {
...
-            let x_a = {
-                let cell =
-                    region.assign_advice_from_constant(|| "fixed x_q", config.x_a, offset, x_q)?;
-                CellValue::new(cell, Some(x_q))
-            };
+            let x_a =
+                region.assign_advice_from_constant(|| "fixed x_q", config.x_a, offset, x_q)?;
 
             x_a.into()
         };
```
Comment (daira) [inline src/circuit/gadget/sinsemilla/chip/hash_to_point.rs:60]:
```suggestion
        let mut x_a: X<pallas::Base> =
            region.assign_advice_from_constant(|| "fixed x_q", config.x_a, offset, x_q)?.into();
```

File: src/circuit/gadget/sinsemilla/chip/hash_to_point.rs [hunk 8]
```diff
@@ -46,16 +53,13 @@ impl SinsemillaChip {
...
```
Comment (daira) [inline src/circuit/gadget/sinsemilla/chip/hash_to_point.rs:60]:
```
        let mut x_a: X<pallas::Base> =
            region.assign_advice_from_constant(|| "fixed x_q", config.x_a, offset, x_q)?.into();
```

File: src/circuit/gadget/poseidon.rs (hunk 1 above)
Comment (str4d) [reply src/circuit/gadget/poseidon.rs:152]: Because of the removal of the `Copy` bound; `let x = [init_val; N];` requires that `init_val` can be copied.

File: src/circuit/gadget/ecc/chip/add_incomplete.rs (hunk 4 above)
Comment (str4d) [reply src/circuit/gadget/ecc/chip/add_incomplete.rs:140]: I was trying to keep the diff as small as possible 😅 but yes, I can go through and clean up variable names.

File: src/circuit/gadget/ecc/chip/mul/overflow.rs [hunk 9]
```diff
@@ -135,16 +135,10 @@ impl Config {
...
-                copy(
-                    &mut region,
-                    || "copy z_130",
-                    self.advices[0],
-                    offset + 1,
-                    &*zs[130],
-                )?;
+                zs[130].copy_advice(|| "copy z_130", &mut region, self.advices[0], offset + 1)?;
 
                 // Witness η = inv0(z_130), where inv0(x) = 0 if x = 0, 1/x otherwise
```
Comment (daira) [inline src/circuit/gadget/ecc/chip/mul/overflow.rs:144]: (For future reference) It's a common pattern to use inv0 to implement nonzero constraints. I'd really like to have an easy way to defer these inverses so that they can be *automatically* batched. To do that, the chip code shouldn't compute the inverse itself, it should call some method on the region that assigns an inverse. Then any use of an inverted value triggers the batch inversion.

File: src/circuit/gadget/sinsemilla/commit_ivk.rs [hunk 10]
```diff
@@ -474,28 +474,24 @@ impl CommitIvkConfig {
                 {
                     let offset = 0;
                     // Copy in `ak`
```
Comment (daira) [inline src/circuit/gadget/sinsemilla/commit_ivk.rs:488]: Consider deleting the "Copy in ..." and "Witness ..." comments in this method, which don't say anything that isn't obvious from the code.

File: src/circuit/gadget/ecc/chip/mul/overflow.rs (hunk 9 above)
Comment (str4d) [reply src/circuit/gadget/ecc/chip/mul/overflow.rs:144]: We already have support for this during witness assignment: that's what `Assigned<F>` represents.

File: src/circuit/gadget/ecc/chip/mul_fixed/short.rs [hunk 11]
```diff
@@ -73,7 +73,10 @@ impl Config {
         &self,
         region: &mut Region<'_, pallas::Base>,
         offset: usize,
-        magnitude_sign: (CellValue<pallas::Base>, CellValue<pallas::Base>),
+        magnitude_sign: (
+            AssignedCell<pallas::Base, pallas::Base>,
+            AssignedCell<pallas::Base, pallas::Base>,
```
Comment (daira) [inline src/circuit/gadget/ecc/chip/mul_fixed/short.rs:78]: These could have a narrower value type.

File: src/circuit/gadget/ecc/chip.rs [hunk 12]
```diff
@@ -349,9 +350,11 @@ pub struct EccScalarFixed {
 /// k_21 must be a single bit, i.e. 0 or 1.
 #[derive(Clone, Debug)]
 pub struct EccScalarFixedShort {
-    magnitude: CellValue<pallas::Base>,
-    sign: CellValue<pallas::Base>,
-    running_sum: ArrayVec<CellValue<pallas::Base>, { constants::NUM_WINDOWS_SHORT + 1 }>,
+    magnitude: AssignedCell<pallas::Base, pallas::Base>,
```
Comment (daira) [inline src/circuit/gadget/ecc/chip.rs:353]:
```suggestion
    // TODO: Make V a `u64`.
    magnitude: AssignedCell<pallas::Base, pallas::Base>,
```

File: src/circuit/gadget/ecc/chip.rs [hunk 13]
```diff
@@ -365,23 +368,24 @@ pub struct EccScalarFixedShort {
...
+    fn base_field_elem(&self) -> AssignedCell<pallas::Base, pallas::Base> {
+        self.base_field_elem.clone()
     }
 }
 
 impl EccInstructions<pallas::Affine> for EccChip {
     type ScalarFixed = EccScalarFixed;
     type ScalarFixedShort = EccScalarFixedShort;
-    type ScalarVar = CellValue<pallas::Base>;
+    // TODO: V = pallas::Scalar?
```
Comment (daira) [inline src/circuit/gadget/ecc/chip.rs:384]: No; ScalarVar is used for variable-base scalar multiplication where the scalar is represented as a base element in F<sub>p</sub>, and the implementation assumes that p < q. So `V = pallas::Base` is correct.

```suggestion
```

File: src/circuit/gadget/ecc/chip.rs [hunk 14]
```diff
@@ -491,7 +495,11 @@ impl EccInstructions<pallas::Affine> for EccChip {
     fn mul_fixed_short(
         &self,
         layouter: &mut impl Layouter<pallas::Base>,
-        magnitude_sign: (CellValue<pallas::Base>, CellValue<pallas::Base>),
+        // TODO: Return AssignedCell<Sign, pallas::Base>
```
Comment (daira) [inline src/circuit/gadget/ecc/chip.rs:498]:
```suggestion
        // TODO: Return (AssignedCell<u64, pallas::Base>, AssignedCell<Sign, pallas::Base>)
```

File: src/circuit/gadget/ecc/chip.rs [hunk 15]
```diff
@@ -491,7 +495,11 @@ impl EccInstructions<pallas::Affine> for EccChip {
...
+        magnitude_sign: (
+            AssignedCell<pallas::Base, pallas::Base>,
+            AssignedCell<pallas::Base, pallas::Base>,
+        ),
```
Comment (daira) [inline src/circuit/gadget/ecc/chip.rs:502]: I suggest making `MagnitudeSign` a typedef for
`(AssignedCell<pallas::Base, pallas::Base>, AssignedCell<pallas::Base, pallas::Base>)` now, then changing it to `(AssignedCell<u64, pallas::Base>, AssignedCell<Sign, pallas::Base>)` later.

```suggestion
        magnitude_sign: MagnitudeSign,
```

File: src/circuit/gadget/ecc/chip/mul_fixed/short.rs [hunk 16]
```diff
@@ -75,15 +73,18 @@ impl Config {
...
+        ),
```
Comment (daira) [inline src/circuit/gadget/ecc/chip/mul_fixed/short.rs:79]:
```suggestion
        magnitude_sign: MagnitudeSign,
```

File: src/circuit/gadget/ecc/chip/mul_fixed/short.rs [hunk 17]
```diff
@@ -99,7 +100,10 @@ impl Config {
     pub fn assign(
         &self,
         mut layouter: impl Layouter<pallas::Base>,
-        magnitude_sign: (CellValue<pallas::Base>, CellValue<pallas::Base>),
+        magnitude_sign: (
+            AssignedCell<pallas::Base, pallas::Base>,
+            AssignedCell<pallas::Base, pallas::Base>,
+        ),
```
Comment (daira) [inline src/circuit/gadget/ecc/chip/mul_fixed/short.rs:106]:
```suggestion
        magnitude_sign: MagnitudeSign,
```

File: src/circuit/gadget/ecc/chip/mul_fixed/short.rs [hunk 18]
```diff
@@ -267,7 +264,13 @@ pub mod tests {
...
             magnitude: pallas::Base,
             sign: pallas::Base,
-        ) -> Result<(CellValue<pallas::Base>, CellValue<pallas::Base>), Error> {
+        ) -> Result<
+            (
+                AssignedCell<pallas::Base, pallas::Base>,
+                AssignedCell<pallas::Base, pallas::Base>,
+            ),
+            Error,
+        > {
```
Comment (daira) [inline src/circuit/gadget/ecc/chip/mul_fixed/short.rs:273]:
```suggestion
        ) -> Result<MagnitudeSign, Error> {
```

File: src/circuit/gadget/ecc/chip/witness_point.rs [hunk 19]
```diff
@@ -70,7 +70,13 @@ impl Config {
...
         offset: usize,
         region: &mut Region<'_, pallas::Base>,
-    ) -> Result<(CellValue<pallas::Base>, CellValue<pallas::Base>), Error> {
+    ) -> Result<
+        (
+            AssignedCell<pallas::Base, pallas::Base>,
+            AssignedCell<pallas::Base, pallas::Base>,
+        ),
+        Error,
+    > {
```
Comment (daira) [inline src/circuit/gadget/ecc/chip/witness_point.rs:79]:
```suggestion
    ) -> Result<MagnitudeSign, Error> {
```

File: src/circuit/gadget/sinsemilla/chip.rs [hunk 20]
```diff
@@ -239,14 +238,14 @@ impl SinsemillaChip {
 impl SinsemillaInstructions<pallas::Affine, { sinsemilla::K }, { sinsemilla::C }>
     for SinsemillaChip
 {
-    type CellValue = CellValue<pallas::Base>;
+    type CellValue = AssignedCell<pallas::Base, pallas::Base>;
```
Comment (daira) [inline src/circuit/gadget/sinsemilla/chip.rs:241]: Consider renaming this.

**Review by daira — APPROVED** [review]
utACK with comments/suggestions.

File: src/circuit/gadget/ecc/chip/mul/overflow.rs (hunk 9 above)
Comment (therealyingtong) [reply src/circuit/gadget/ecc/chip/mul/overflow.rs:144]: So instead of (for example) `z_130.invert().unwrap()` should we do `(pallas::Base::one(), z_130).into()`?

File: src/circuit/gadget/ecc/chip/add.rs (hunk 2 above)
Comment (therealyingtong) [reply src/circuit/gadget/ecc/chip/add.rs:317]: Could be because `-y_p` is dereferenced?

File: src/circuit/gadget/ecc/chip/mul/overflow.rs (hunk 9 above)
Comment (str4d) [reply src/circuit/gadget/ecc/chip/mul/overflow.rs:144]: `Assigned::from(z_130).invert()` would be clearer.

**Comment by str4d** [comment]
[Rebased on #237](https://github.com/zcash/orchard/compare/cd1c8f1c45ab6880986cdf64924225dec22394a5..d12fc94839c031665afa614f56ea4c7fdabccc6a) to fix merge conflicts (which were all in `use` statements). I'll address @daira's comments next.

File: src/circuit/gadget/ecc/chip/witness_point.rs (hunk 19 above)
Comment (str4d) [reply src/circuit/gadget/ecc/chip/witness_point.rs:85]: This one is a point, not a magnitude-sign. Which makes your point :P

**Comment by str4d** [comment]
[Force-pushed](https://github.com/zcash/orchard/compare/d12fc94839c031665afa614f56ea4c7fdabccc6a..bacdf67428f8535e0cfcdfde7529fa440dad6496) to remove an incorrect TODO comment.

File: src/circuit/gadget/ecc/chip/mul/overflow.rs (hunk 9 above)
Comment (str4d) [reply src/circuit/gadget/ecc/chip/mul/overflow.rs:148]: I tried this locally, and discovered a bug in the `AssignedCell` `halo2` PR! Fixed here: https://github.com/zcash/halo2/pull/417 (which I'll immediately start depending on in this PR).

Comment (str4d) [reply src/circuit/gadget/ecc/chip/mul/overflow.rs:148]: Hmm, actually this gets more complex to migrate to in all but a single case, so I'll do this in a separate PR.

File: src/circuit/gadget/ecc/chip/add_incomplete.rs (hunk 4 above)
Comment (str4d) [reply src/circuit/gadget/ecc/chip/add_incomplete.rs:153]: Decided to not do this in this PR, it's already large enough. This can be done post-migration to `halo2_gadgets`.

File: src/circuit/gadget/ecc/chip/mul_fixed/full_width.rs [hunk 21]
```diff
@@ -294,7 +295,9 @@ pub mod tests {
         {
             let scalar_fixed = pallas::Scalar::zero();
             let (result, _) = base.mul(layouter.namespace(|| "mul by zero"), Some(scalar_fixed))?;
-            assert!(result.inner().is_identity().unwrap());
+            if let Some(is_identity) = result.inner().is_identity() {
+                assert!(is_identity);
+            }
```
Comment (ebfull) [inline src/circuit/gadget/ecc/chip/mul_fixed/full_width.rs:300]: We should panic on `else` here right?

**Review by ebfull — APPROVED** [review]

File: src/circuit/gadget/ecc/chip/mul_fixed/full_width.rs (hunk 21 above)
Comment (str4d) [reply src/circuit/gadget/ecc/chip/mul_fixed/full_width.rs:300]: No; that is what the prior behaviour was, and it prevented the test circuit from being used in non-witness contexts like keygen or dev tooling. We can only enforce this assertion if we have a witness to enforce it on.

File: src/circuit/gadget/ecc/chip/add.rs (hunk 2 above)
Comment (daira) [reply src/circuit/gadget/ecc/chip/add.rs:336]: Yes, with some wrinkles. The LHS and RHS of `==` are [place expression contexts](https://doc.rust-lang.org/reference/expressions.html#place-expressions-and-value-expressions), which means they are implicitly borrowed. But because *both* sides are implicitly borrowed, we end up with `&y_q == &-y_p`, in which the reference levels are still mismatched (although the error message says `no implementation for `&pasta_curves::Fp == pasta_curves::Fp` and doesn't mention the implicit borrowing, which I may report as a bug).
